cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -208,26 +212,50 @@ void resumeTasks(Consumer<ProgramRunId> taskCleanup) {
           continue;
       }
 
-      LOG.info("Resuming provisioning task for run {} of type {} in state {}.", provisioningTaskInfo.getProgramRunId(),
-               provisioningTaskInfo.getProvisioningOp().getType(),
-               provisioningTaskInfo.getProvisioningOp().getStatus());
-      taskExecutor.submit(taskKey, task);
+      // the actual task still needs to be run for cleanup to happen, but avoid logging a confusing
+      // message about resuming a task that is in cancelled state.
+      if (provisioningOp.getStatus() != ProvisioningOp.Status.CANCELLED) {
+        LOG.info("Resuming provisioning task for run {} of type {} in state {}.",
+                 provisioningTaskInfo.getProgramRunId(), provisioningTaskInfo.getProvisioningOp().getType(),
+                 provisioningTaskInfo.getProvisioningOp().getStatus());
+      }
+      task.run();
     }
   }
 
   /**
-   * Cancel any provisioning operation currently taking place for the program run
+   * Cancel the provision task for the program run. If there is no task or the task could not be cancelled before it
+   * completed, an empty optional will be returned. If the task was cancelled, the state of the cancelled task will
+   * be returned, and a message will be sent to mark the program run state as killed.
+   *
+   * @param programRunId the program run
+   * @return the state of the task if it was cancelled
+   */
+  public Optional<ProvisioningTaskInfo> cancelProvisionTask(ProgramRunId programRunId) {
+    ProvisioningTaskKey taskKey = new ProvisioningTaskKey(programRunId, ProvisioningOp.Type.PROVISION);
+    return cancelTask(taskKey, programStateWriter::killed);
+  }
+
+  /**
+   * Cancel the deprovision task for the program run. If there is no task or the task could not be cancelled before it
```

Comment:
Just out of curiosity - what happens to the cluster state when you cancel a deprovision task, after it has tore down some stuff?